typescript-requiring-type-checking.js

```diff
@@ -15,23 +15,23 @@ module.exports = {
         /*
```

Comment:
Discussion for recommended rule: `'@typescript-eslint/restrict-template-expressions': 'error'`

Reply:
I noticed that [restrict-template-expressions](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/restrict-template-expressions.md) would trigger in the default case of the following code, saying "Invalid type "never" of template literal expression":
```ts
enum Foo { First, Second }

someFunc(x: Foo) {
    switch (x) {
        case Foo.First: return 'first';
        case Foo.Second: return 'second';
        default: throw new Error(`Invalid enum value ${x}`);
    }
}
```
What would be our recommended fix for this error?

Reply:
We should check if we have a rule that ensures `switch` statements are fully covering every enum value. If so then the recommendation would be not to provide a `default` case.

In cases where we don't know all the enum values (if coming from an external library) then it would be appropriate to cast the value in the template literal to tell the compiler what's going on (if necessary...in that case it might not be `never` type)

Reply:
[switch-exhaustiveness-check](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md) will check enums and enum like types in switch statements. We already enabled [default-case](https://eslint.org/docs/rules/default-case), so excluding the default case will require a comment.

Reply:
@jattasNI will verify that switch-exhaustiveness-check and default-case play well together. 

The restrict-template-expressions rule that triggered this discussion is fine.

Reply:
**Proposal** Keep `default-case` as `error` but turn off `switch-exhaustiveness-check`.
```
/* Always provide a `default` case in `switch` statements. If the default case is logically unreachable, throw an Error. If the default case should do nothing, add a comment explaining why to indicate that it is an intentional decision.*/
"default-case": "error"

/* This rule is made redundant by `default-case` so it will never fire
"@typescript-eslint/switch-exhaustiveness-check": "off"
```

Here's my rationale:
1. There are some patterns that only `default-case` can detect. For example, when exhaustiveness can't be calculated (switching on a `string` with any possible value), our ruleset should enforce having a `default` case to signal that the developer thought about all possible values.
2. `default-case` is a bit annoying when the compiler and `switch-exhaustiveness-check` detect that the cases are exhaustive because it forces an unnecessary default case. But even though the case is unreachable, it doesn't cause any compiler errors. And it gives us a place to add a `throw` to make it obvious when reading the code that it is unreachable.
3. Once `default-case` is on, `switch-exhaustiveness-check` would never fail because adding a `default` case will always make the `switch` exhaustive. It's not worth the time to enforce it so we should turn it off.

I created an [example](https://github.com/jattasNI/typescript-eslint-sandbox/blob/main/src/index.ts) with all the cases I could think of if you want have a visual aid to think it through.

Reply:
@rajsite this thread is ready to close